File: src/main/java/no/novari/flyt/authorization/user/model/UserEntity.java
package no.novari.flyt.authorization.user.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.hibernate.annotations.NaturalId;

import jakarta.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Getter
@Builder
@Entity
@NoArgsConstructor
@AllArgsConstructor
public class UserEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonIgnore
    @Setter(AccessLevel.NONE)
    private long id;

    @Setter
    @NaturalId
    @Column(nullable = false, unique = true)
    private UUID objectIdentifier;

    @Setter
    private String email;

    @Setter
    private String name;

    @ElementCollection
    @CollectionTable(
            name = "user_entity_source_application_ids",
            joinColumns = @JoinColumn(name = "user_entity_id"),
            uniqueConstraints = @UniqueConstraint(columnNames = {"user_entity_id", "source_application_ids"})
    )
    @Setter
    @Column(name = "source_application_ids")
    @Builder.Default
    private List<Long> sourceApplicationIds = new ArrayList<>();
}

File: src/main/kotlin/no/novari/flyt/authorization/user/model/UserEntity.kt
package no.novari.flyt.authorization.user.model

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.CollectionTable
import jakarta.persistence.Column
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.UniqueConstraint
import no.novari.flyt.audit.entity.AuditedEntity
import org.hibernate.annotations.NaturalId
import org.hibernate.envers.AuditJoinTable
import org.hibernate.envers.AuditTable
import org.hibernate.envers.Audited
import java.util.UUID

@Entity
@Audited
@AuditTable("user_entity_aud")
class UserEntity(
    @NaturalId
    @Column(nullable = false, unique = true)
    var objectIdentifier: UUID? = null,
    var email: String? = null,
    var name: String? = null,
    @Audited
    @AuditJoinTable(name = "user_entity_source_application_ids_aud")
    @ElementCollection
    @CollectionTable(
        name = "user_entity_source_application_ids",
        joinColumns = [JoinColumn(name = "user_entity_id")],
        uniqueConstraints = [UniqueConstraint(columnNames = ["user_entity_id", "source_application_ids"])],
    )
    @Column(name = "source_application_ids")
    var sourceApplicationIds: MutableList<Long> = mutableListOf(),
) : AuditedEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonIgnore
    var id: Long? = null
}
